### PR TITLE
 Visual Indicator of Where Pipeline Rules are applied

### PR DESCRIPTION
**Pull request by diesaulov**
Extending the Pipeline Rule API to return a list of using pipelines for each rule in a paginated response.

Response example:
`GET /api/system/pipelines/rule/paginated?page=1&per_page=10`
```json
{
  "total": 4,
  "page": 1,
  "per_page": 10,
  "count": 4,
  "rules": [
    {
      "title": "debug#1",
      "description": "",
      "source": "rule \"debug#1\"\nwhen \n    true\nthen\n    debug(\"OK#1\");\nend",
      "created_at": "2021-10-24T11:03:56.514Z",
      "modified_at": "2021-10-24T11:03:56.514Z",
      "errors": null,
      "id": "61753d9c2f64de4a2582ea55"
    },
    {
      "title": "debug#2",
      "description": "Debug rule 2",
      "source": "rule \"debug#2\"\nwhen \n    true\nthen\n    debug(\"OK#2\");\nend",
      "created_at": "2021-10-24T11:04:15.778Z",
      "modified_at": "2021-10-24T11:04:15.778Z",
      "errors": null,
      "id": "61753daf2f64de4a2582ea6c"
    },
    {
      "title": "debug#3",
      "description": "",
      "source": "rule \"debug#3\"\nwhen \n    true\nthen\n    debug(\"OK#3\");\nend",
      "created_at": "2021-10-24T11:04:34.330Z",
      "modified_at": "2021-10-24T11:04:34.330Z",
      "errors": null,
      "id": "61753dc22f64de4a2582ea82"
    },
    {
      "title": "debug#4",
      "description": "",
      "source": "rule \"debug#4\"\nwhen \n    true\nthen\n    debug(\"OK#4\");\nend",
      "created_at": "2021-10-24T11:56:57.584Z",
      "modified_at": "2021-10-24T11:56:57.584Z",
      "errors": null,
      "id": "61754a0900b6ae48cce45a32"
    }
  ],
  "context": {
    "used_in_pipelines": {
      "61753daf2f64de4a2582ea6c": [
        {
          "id": "6175495100b6ae48cce45965",
          "title": "Pipeline 1"
        },
        {
          "id": "6175496000b6ae48cce45977",
          "title": "Pipeline 2"
        }
      ],
      "61753d9c2f64de4a2582ea55": [
        {
          "id": "6175495100b6ae48cce45965",
          "title": "Pipeline 1"
        }
      ],
      "61754a0900b6ae48cce45a32": [],
      "61753dc22f64de4a2582ea82": [
        {
          "id": "6175496000b6ae48cce45977",
          "title": "Pipeline 2"
        }
      ]
    }
  }
}
```

Closes https://github.com/Graylog2/graylog2-server/issues/9724

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/db/PipelineDao.java [hunk 1]
```diff
@@ -58,6 +60,10 @@
     @Nullable
     public abstract DateTime modifiedAt();
 
+    public boolean usesRule(String ruleTitle) {
+        return source().contains(f("rule \"%s\"", ruleTitle));
```
[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/db/PipelineDao.java]: https://github.com/Graylog2/graylog2-server/pull/11474/commits/53eaeee52b9460bc56d023b3fa2f5732d1a1ce6f

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/db/mongodb/MongoDbPipelineService.java [hunk 2]
```diff
@@ -89,6 +96,23 @@ public PipelineDao loadByName(String name) throws NotFoundException {
...
 
+    @Override
+    public List<PipelineDao> loadByRules(Set<String> ruleNames) {
+        if (ruleNames.isEmpty()) {
+            return ImmutableList.of();
+        }
+        final DBQuery.Query query = DBQuery.or(
+                ruleNames
+                        .stream()
+                        .map(rn -> DBQuery.regex("source", Pattern.compile(f("rule \"%s\"", rn))))
```
[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/db/mongodb/MongoDbPipelineService.java]: https://github.com/Graylog2/graylog2-server/pull/11474/commits/53eaeee52b9460bc56d023b3fa2f5732d1a1ce6f

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rest/RuleResource.java [hunk 3]
```diff
@@ -190,9 +198,38 @@ public RuleSource parse(@ApiParam(name = "rule", required = true) @NotNull RuleS
...
+                    .filter(p -> p.usesRule(r.title()))
+                    .map(p -> PipelineCompactSource.builder()
+                            .id(p.id())
+                            .title(p.title())
+                            .build())
+                    .collect(Collectors.toList()));
+        });
+        final long took = System.currentTimeMillis() - started;
+        if (took > 500) {
+            log.warn("It takes longer than 500ms to assemble pipeline rule paginated response's context!");
```
[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rest/RuleResource.java]: Correct. Removed: https://github.com/Graylog2/graylog2-server/pull/11474/commits/53eaeee52b9460bc56d023b3fa2f5732d1a1ce6f

File: graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/memory/InMemoryPipelineServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    private final InMemoryPipelineService underTest = new InMemoryPipelineService(new ClusterEventBus());
+
+    @Test
+    public void loadByRules_returnsAllPipelinesUsingProvidedRules() {
+        // given
+        saveNewPipelineDao("Pipeline 1", "pipeline \"Pipeline 1\"\n" +
+                "stage 0 match either\n" +
+                "rule \"debug#1\"\n" +
+                "rule \"debug#2\"\n" +
```
[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/memory/InMemoryPipelineServiceTest.java]: https://github.com/Graylog2/graylog2-server/pull/11474/commits/53eaeee52b9460bc56d023b3fa2f5732d1a1ce6f

**Review by bernd — COMMENTED** [review]
@diesaulov @mpfz0r I think finding the pipeline rules that are referenced in pipelines via regular expressions on the pipeline source code is quite brittle.
We have a parser that can parse the pipeline source code. So using that parser to find out where rules are used would be a way better approach in my opinion.

Are there any reasons not to use the parser?

**Comment by bernd** [comment]
> I guess parsing every pipeline is more expensive than just a regex. However, wouldn't it be possible to access the `PipelineInterpreter.State` and take the required infos from there? That should have all the pipelines parsed already.

@mpfz0r Not sure how fast the database query with regexp is, but even if the regex way is faster, it's still more brittle than parsing it. Since this is not something that runs in a hot loop but only for display purposes, I think it's okay if it's not super fast. :smile:

**Comment by bernd** [comment]
> However, wouldn't it be possible to access the `PipelineInterpreter.State` and take the required infos from there?
> That should have all the pipelines parsed already.

@mpfz0r Aye, that might work as well. :+1: We have to check if we can access it easily.

**Comment by diesaulov** [comment]
@bernd @mpfz0r 

> However, wouldn't it be possible to access the PipelineInterpreter.State and take the required infos from there?
> That should have all the pipelines parsed already.

Intuitively pulling the processing state as a dependency into the pipeline / rule persistence layer doesn't seem to me a good idea. Moreover it will introduce a circular dependency.

If we don't expect thousands of the pipelines to be parsed, then we should be good to simply load and parse all (though at least basic performance tests should be conducted)

**Comment by diesaulov** [comment]
@mpfz0r @bernd that should be it: https://github.com/Graylog2/graylog2-server/pull/11474/commits/12b3717c0e2fb06c227de119b13d905ee8b99c5d

**Review by mpfz0r — CHANGES_REQUESTED** [review]
One last thing, rest looks good

File: graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/DbPipelineServiceBaseTest.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        final PipelineServiceHelper pipelineServiceHelper = new PipelineServiceHelper(pipelineRuleParser);
+
+        underTest = createPipelineService(pipelineServiceHelper);
+    }
+
+    abstract protected PipelineService createPipelineService(PipelineServiceHelper pipelineServiceHelper);
+
+    @Test
+    public void loadByRules_returnsAllPipelinesUsingProvidedRules() {
+        // given
```
[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/DbPipelineServiceBaseTest.java:56]: Why would we store a non-parseable pipeline in the database?

[the remark this answers is not in the document]
Comment (mpfz0r) [reply graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/DbPipelineServiceBaseTest.java:56]: Just for good measure. 
I think we didn't' always had the validation in place that prevents non-parseable rules from being saved

[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/DbPipelineServiceBaseTest.java:56]: Alright, I see. This sort of defensive parsing is not something new in the code base...

[the remark this answers is not in the document]
Comment (diesaulov) [reply graylog2-server/src/test/java/org/graylog/plugins/pipelineprocessor/db/DbPipelineServiceBaseTest.java:56]: https://github.com/Graylog2/graylog2-server/pull/11474/commits/71e6997726a2c3471af9370fcd5e1bb87b848959

**Review by mpfz0r — APPROVED** [review]
LGTM :+1: 
@bernd any other comments from your side?

**Comment by bernd** [comment]
>  @bernd any other comments from your side?
@mpfz0r Nope, thank you! :+1:

**Pull request merged**